### PR TITLE
BananaPi BPI-R4 board: add support for 2.5GE PoE variant

### PR DESCRIPTION
**Pull request by sch-m**
This adds support for the BPI-R4 board-variant with the internal 2.5G phy.

For this I have introduced new DTS overlays, analogous to the scheme of the MT7988a RFB.
A new uboot environment variable `bootconf_eth1` makes it possible to switch between the variants.

Remaining "problems":
1. Moving the eth1 configuration to a new overlay and the new associated bootconf variable in the uboot means that the system is no longer compatible with previous uboot environments.

@dangowrt What is your opinion on this or do you have any great ideas on how to remain backwards compatible?

2. We need to change the polarity for led0 of the i2p5g phy to `active-high` to get it working. We should add a dts property to select the polarity of the leds.

### CODE REVIEW

**Comment by dangowrt** [comment]
I don't think the use of DT overlays is appropriate in this case, because it's really a different board, it should be it's own U-Boot build and a separate image build in OpenWrt.

**Comment by sch-m** [comment]
It's not really a different board, just a different board variant.

I also think that a separate uboot build and image build could be justified here, but due to the different boot modes (SD, eMMC, SPI-NAND) this results in 3 further uboot builds that only differ from the previous uboot in this one small point.

Irrespective of this, the approach using DT overlays would have the advantage that no redundant DTS files would have to be created, but only the actual differences would be maintained in different overlays. This still applies, even if you create a new image for this variant. I thought that is exactly the use case for overlays.

**Comment by dangowrt** [comment]
> 3 further uboot builds that only differ from the previous uboot in this one small point.

Sadly that's how it is. But it's true that violating the rule in this case would spare some buildbot resources...

> Irrespective of this, the approach using DT overlays would have the advantage that no redundant DTS files would have to be created

You can avoid that be moving the common parts into a DTSI, which would be the appropriate thing to do in this case.

> I thought that is exactly the use case for overlays.

In OpenWrt we use overlays for two things:
 - boards which have multiple hardware or software **user-modifiable** configurations (such as DIP switches changing the boot device, or, in case of the BPi-R64, having either a SATA host or a 2nd PCIe host)
 - additional peripherals (such as the RTC battery to the BPi-R4, or even ReSpeaker soundcard to BPi-R3, or any Pi HATs and stuff like that)

**Comment by sch-m** [comment]
> You can avoid that be moving the common parts into a DTSI, which would be the appropriate thing to do in this case.

Well, then I'll move the common code from `mt7988a-bananapi-bpi-r4.dts` to `mt7988a-bananapi-bpi-r4.dtsi`, keep the sfp2 related nodes in `mt7988a-bananapi-bpi-r4.dts` (to stay compatible with existing uboot) and create a new ``mt7988a-bananapi-bpi-r4-eth1-i2p5g-phy.dts` for the new variant.

Then also create new uboot and image variants with the `-eth1-i2p5g-phy` postfix.

OK?

**Comment by sch-m** [comment]
Hmmm... it should also be possible to put the sfp part into a `mt7988a-bananapi-bpi-r4-eth1-sfp.dts`, because only `DEVICE_DTS_CONFIG := config-mt7988a-bananapi-bpi-r4` is relevant for the uboot.

In this way it should also be possible to have 2 device sections in filogic.mk which only differ in `DEVICE_DTS` and thus use the same uboot images.

**Comment by openips** [comment]
nice

**Comment by sch-m** [comment]
> Is the PoE/i2p5g variant actually sold anywhere?

We have ordered custom boards from Sinovoip with the RJ45 Jack mounted instead of the SFP cage. No PoE.
As even with PoE the differences in the the DTS are related to the 2.5G ETH and not to the PoE feature, I think calling the new board `banana-bpi-r4-2p5g` is more suitable.

**Comment by sch-m** [comment]
@dangowrt I have updated the PR. It now uses a dtsi and new dts files instead of additional overlays. Furthermore, additional uboot images are created and there is a separate device in filogic.mk.

**Review by openips — APPROVED** [review]

**Review by dangowrt — APPROVED** [review]
Looks good to me now. I'd still prefer the name being `-poe` instead of `-2p5ge` because it's closer to the marketing name, but I won't fight for that. Why did you chose to order those custom boards instead of just adding an RJ-45 SFP module if anyway PoE was not part of the story?

Also, setting up LED polarity from DT can wait until we have more boards using the internal 2.5GE PHY.

**Comment by sch-m** [comment]
> Why did you chose to order those custom boards instead of just adding an RJ-45 SFP module if anyway PoE was not part of the story?

Well, I think thats a matter of costs. We are planning to create a product based on the BPI-R4 and our product management has opted for this variant.

**Comment by sch-m** [comment]
> I'd still prefer the name being -poe instead of -2p5ge because it's closer to the marketing name, but I won't fight for that.

Even if I personally find the `-2p5ge` more appropriate, I can change it to `-poe`. As we will probably be connecting additional extensions (e.g. via i2c etc.), I will have to create another DTS for our variant anyway.

**Comment by sch-m** [comment]
I have just noticed that I still have to add the board_name `bananapi,bpi-r4-2p5ge` or `bananapi,bpi-r4-poe` to various scripts (01_leds, 02_network, upgrade script, uboot-env etc.).

**Comment by mkowalski** [comment]
Hey @sch-m, as I am reading through the history of comments, it seems like it was developed based on a custom build by Sinovoip and not the board they are selling today? Is this correct?

I am asking because I seem to have lots of issues getting the 2.5GbE interface working on the latest OpenWrt even though it works with the outdated "factory build".

There is some ongoing discussion in https://forum.banana-pi.org/t/bpi-r4-no-link-on-2-5g-interface but I would appreciate your feedback as you seem to have had access to the board and got it working.

**TLDR: latest openwrt build can never bring the link up on the 2.5GbE interface**

**Comment by sch-m** [comment]
Hi @mkowalski!

It's been a while since I tested with the board.
Unfortunately I can't try it out today, but on Friday I will install a current master and test it.
I'll let you know what the results are.

**Comment by sch-m** [comment]
@mkowalski 
I've testet my board with a custom build based on the openwrt commit https://github.com/openwrt/openwrt/commit/2422dddf7a9a383214fcfa671f18c8cbb463163a. The 2.5G link comes up and communication is possible.

**Comment by sch-m** [comment]
I've also picked the `openwrt-mediatek-filogic-bananapi_bpi-r4-poe-sdcard.img.gz` from [here](https://downloads.openwrt.org/snapshots/targets/mediatek/filogic/) (sha256sum: 2a2dbf6372e0fab0e8266a7434198c7e165551c9a7594e351f7b8e6209d3f4c2, date: Thu Oct 3 21:58:03 2024) and installed it on a clean sd card.
Same result: The 2.5G port works as expected.

**Comment by mkowalski** [comment]
Thanks for the confirmation. Probably my particular board is dodged then
